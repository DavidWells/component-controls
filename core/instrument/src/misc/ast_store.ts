import * as fs from 'fs';
import * as parser from '@babel/parser';
import { File } from '@babel/types';
import traverse from '@babel/traverse';
import { ImportTypes } from '@component-controls/core';
import { traverseImports } from '../babel/extract-imports';
import { traverseExports, ExportTypes } from '../babel/extract-exports';

type CacheProps = {
  ast: File;
  source: string;
  imports?: ImportTypes;
  exports?: ExportTypes;
};
const astCache: Record<string, CacheProps> = {};

export const parseFile = (
  filePath: string,
  options?: parser.ParserOptions,
  sourceCode?: string,
): CacheProps => {
  if (astCache[filePath]) {
    return astCache[filePath];
  }
  const source = sourceCode || fs.readFileSync(filePath, 'utf8');
  const ast = parser.parse(source, options);
  astCache[filePath] = { ast, source };
  return astCache[filePath];
};

export const parseImports = (
  fileName: string,
  options?: parser.ParserOptions,
): ImportTypes => {
  const cache = parseFile(fileName, options);
  if (!cache.imports) {
    cache.imports = {};
    traverse(cache.ast, traverseImports(cache.imports));
  }
  return cache.imports;
};

export const parseExports = (
  fileName: string,
  options?: parser.ParserOptions,
): ExportTypes => {
  const cache = parseFile(fileName, options);
  if (!cache.exports) {
    cache.exports = { named: {} };
    traverse(cache.ast, traverseExports(cache.exports));
  }
  return cache.exports;
};
