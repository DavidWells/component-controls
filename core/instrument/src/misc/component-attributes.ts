import { Story, Document } from '@component-controls/core';

export const componentsFromParams = (
  element: (Document | Story) & { of?: string },
): string[] => {
  const result = [];
  let { component } = element;

  if (typeof component === 'string') {
    result.push(component);
  }
  const { of: componentShorthand } = element;
  if (typeof componentShorthand === 'string') {
    result.push(componentShorthand);
  }
  let { subcomponents } = element;
  if (typeof subcomponents === 'string') {
    result.push(subcomponents);
  }
  if (typeof subcomponents === 'object') {
    //@ts-ignore
    Object.keys(subcomponents).forEach(key => result.push(subcomponents[key]));
  }
  if (typeof componentShorthand === 'object') {
    Object.keys(componentShorthand).forEach(key =>
      result.push(componentShorthand[key]),
    );
  }

  return result;
};
