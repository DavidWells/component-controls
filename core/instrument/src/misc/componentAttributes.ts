import {
  Story,
  StoriesKind,
  StoryParameters,
} from '@component-controls/specification';

export const componentsFromParams = (
  element: StoriesKind | Story | StoryParameters,
): string[] => {
  const result = [];
  let { component } = element;
  if (!component && element.parameters) {
    ({ component } = element.parameters);
  }
  if (typeof component === 'string') {
    result.push(component);
  }
  let { subcomponents } = element;
  if (!subcomponents && element.parameters) {
    ({ subcomponents } = element.parameters);
  }
  if (typeof subcomponents === 'string') {
    result.push(subcomponents);
  }
  if (subcomponents) {
    console.log(subcomponents);
  }
  return result;
};
