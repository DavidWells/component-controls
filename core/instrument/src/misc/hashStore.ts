import { createHash } from 'crypto';

export const hashStoreId = (name: string): string =>
  createHash('md5')
    .update(name)
    .digest('hex');
