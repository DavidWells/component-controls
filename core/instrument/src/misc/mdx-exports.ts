import { MDXExportType, MDXExportTypes } from '../types';
import { stringifyObject } from './stringify-object';

const mdxPropertiesExport = (exportType: MDXExportType): string | undefined => {
  return exportType ? stringifyObject(exportType.story) : undefined;
};

const mdxFunctionExport = (
  name: string,
  exportType: MDXExportType,
): string | undefined => {
  return exportType && exportType.render
    ? `export const ${name} = ${exportType.render}`
    : undefined;
};

export const extractStoryExports = (exports?: MDXExportTypes): string => {
  if (exports) {
    const exportNames = Object.keys(exports);
    if (exportNames.length) {
      let defaultExportCode = '';
      if (exports.default) {
        const expCode = mdxPropertiesExport(exports.default);
        defaultExportCode = `
        const dmsPageExport = MDXContent;
        dmsPageExport.MDXPage = MDXContent;

        export default ${
          expCode
            ? `Object.assign(dmsPageExport, ${expCode});`
            : 'dmsPageExport;'
        }
`;
      }

      let storiesExports: string[] = [];
      const expStories = Object.keys(exports).filter(id => id !== 'default');
      if (expStories.length) {
        for (const exportStory of expStories) {
          const expFn = mdxFunctionExport(exportStory, exports[exportStory]);
          if (expFn) {
            storiesExports.push(expFn);
          }
          const expCode = mdxPropertiesExport(exports[exportStory]);
          if (expCode) {
            storiesExports.push(`${exportStory}.story = ${expCode}
             `);
          }
        }
      }
      return `${defaultExportCode}\n${storiesExports.join('\n')}`;
    }
  }
  return `
const dmsPageExport = MDXContent;
dmsPageExport.MDXPage = MDXContent;
export default  dmsPageExport;
`;
};
