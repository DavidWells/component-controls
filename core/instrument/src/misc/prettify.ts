import prettier from 'prettier';
import parserBabel from 'prettier/parser-babylon';
import { PrettierOptions } from '../types';

export const prettify = async (
  code: string,
  filePath: string,
  options: PrettierOptions | false,
): Promise<string> => {
  if (options !== false) {
    const { resolveConfigOptions, ...otherOptions } = options || {};
    let allPrettierOptions: prettier.Options | undefined = otherOptions as any;
    if (filePath) {
      const userOptions = await prettier.resolveConfig(
        filePath,
        resolveConfigOptions,
      );
      allPrettierOptions = { ...userOptions, ...allPrettierOptions };
    }
    try {
      return prettier.format(code, {
        parser: 'typescript',
        plugins: [parserBabel],
        ...allPrettierOptions,
      });
    } catch (e) {
      return code;
    }
  } else {
    return code || '';
  }
};
