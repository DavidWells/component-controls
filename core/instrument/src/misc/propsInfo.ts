import * as fs from 'fs';
import * as path from 'path';
import * as os from 'os';
import { createHash } from 'crypto';
import findCacheDir from 'find-cache-dir';
import {
  ComponentInfo,
  PropsInfoExtractorFunction,
} from '@component-controls/specification';
import { PropsLoaderConfig } from '../types';

export const propsInfo = async (
  options: PropsLoaderConfig[],
  filePath: string,
  componentName?: string,
  source?: string,
): Promise<ComponentInfo | undefined> => {
  const cacheFolder =
    findCacheDir({ name: 'component-controls-props-info' }) || os.tmpdir();

  //create cache folder if it doesnt exist
  if (!fs.existsSync(cacheFolder)) {
    fs.mkdirSync(cacheFolder, { recursive: true });
  }
  const cachedFileName = path.join(
    cacheFolder,
    createHash('md5')
      .update(`${filePath}-${componentName}`)
      .digest('hex'),
  );
  if (fs.existsSync(cachedFileName)) {
    const cacheStats = fs.statSync(cachedFileName);
    const fileStats = fs.statSync(filePath);
    if (cacheStats.mtime.getTime() >= fileStats.mtime.getTime()) {
      const fileData = fs.readFileSync(cachedFileName, 'utf8');
      const json = JSON.parse(fileData);
      return Object.keys(json).length ? json : undefined;
    }
  }
  let result: ComponentInfo | undefined = undefined;
  const loaders = options.filter(loader => {
    const include = Array.isArray(loader.test)
      ? loader.test
      : loader.test
      ? [loader.test]
      : undefined;
    const exclude = Array.isArray(loader.exclude)
      ? loader.exclude
      : loader.exclude
      ? [loader.exclude]
      : undefined;
    return (
      include &&
      include.some(mask => filePath.match(mask)) &&
      (!exclude || !exclude.some(mask => filePath.match(mask)))
    );
  });

  if (loaders.length > 1) {
    console.error(`Multiple propsloaders found for file ${filePath}`);
  }
  const propsLoaderName = loaders.length === 1 ? loaders[0] : undefined;
  if (propsLoaderName) {
    const { run } = require(propsLoaderName.name);
    if (run) {
      const propsLoader: PropsInfoExtractorFunction = run(
        propsLoaderName.options,
      );
      result = await propsLoader(filePath, componentName, source);
    }
  }
  fs.writeFileSync(cachedFileName, JSON.stringify(result || {}));
  return result;
};
