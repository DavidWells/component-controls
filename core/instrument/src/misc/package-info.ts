import * as fs from 'fs';
import * as path from 'path';
//@ts-ignore
import readJson from 'read-package-json';
import hostedGitInfo from 'hosted-git-info';
//@ts-ignore
import parseRepositoryURL from '@hutson/parse-repository-url';
import { PackageInfo } from '@component-controls/specification';
import { hashStoreId } from './hashStore';
import { PackageInfoOptions } from '../types';

const traverseFolder = (
  filePath: string,
  levels: number,
  packageJsonName: string,
): string | null => {
  var files = fs.readdirSync(filePath);
  if (levels === 0) {
    return null;
  }
  const pckg = files.find(file => file === packageJsonName);
  if (pckg) {
    return path.resolve(filePath, pckg);
  }
  return traverseFolder(
    path.resolve(filePath, '..'),
    levels - 1,
    packageJsonName,
  );
};

const getPackageJson = async (
  filePath?: string,
  opts?: PackageInfoOptions | false,
): Promise<{
  fileName: string;
  packageJSON: { [key: string]: any };
} | null> => {
  return new Promise((resolve, reject) => {
    if (!filePath) {
      return resolve(null);
    }
    if (opts === false) {
      return resolve(null);
    }
    const fileName = traverseFolder(
      path.dirname(filePath),
      opts?.maxLevels || 10,
      opts?.packageJsonName || 'package.json',
    );
    if (!fileName) {
      return resolve(null);
    }
    readJson(fileName, console.error, false, (err: string, data: object) => {
      if (err) {
        return reject(err);
      }
      return resolve({ fileName, packageJSON: data });
    });
  });
};

export interface PackageInfoReturnType {
  browse: string;
}

export const packageInfo = async (
  filePath?: string,
  opts?: PackageInfoOptions | false,
): Promise<PackageInfo | undefined> => {
  if (filePath) {
    const { fileName, packageJSON } =
      (await getPackageJson(filePath, opts)) || {};
    if (fileName && packageJSON) {
      const repositoryURL =
        typeof packageJSON.repository === 'string'
          ? packageJSON.repository
          : packageJSON.repository && packageJSON.repository.url;
      if (repositoryURL) {
        const templates =
          hostedGitInfo.fromUrl(repositoryURL) ||
          parseRepositoryURL(repositoryURL);
        const fileRelativePath = `${
          packageJSON.repository && packageJSON.repository.directory
            ? `${packageJSON.repository.directory}/`
            : ''
        }${path.relative(path.dirname(fileName), filePath)}`;
        const fillTemplate = (template: string): string => {
          return template
            .replace('auth@', templates.auth || '')
            .replace('{domain}', templates.domain || '')
            .replace('{user}', templates.user || '')
            .replace('{project}', templates.project || '')
            .replace('{#fragment}', '')
            .replace('{path}', fileRelativePath)
            .replace('{treepath}', templates.treepath || '')
            .replace(
              '{/tree/committish}',
              `/${templates.treepath}/${templates.committish || 'master'}`,
            )
            .replace('{committish}', templates.committish || 'master');
        };
        const {
          name,
          version,
          dependencies,
          devDependencies,
          peerDependencies,
        } = packageJSON;
        const result: PackageInfo = {
          fileHash: hashStoreId(fileName),
          name,
          version,
          repository: {},
          dependencies,
          devDependencies,
          peerDependencies,
        };
        const { storeBrowseLink, storeDocsLink, storeIssuesLink } = opts || {};
        if (storeBrowseLink) {
          result.repository.browse = fillTemplate(templates.browsefiletemplate);
        }
        if (storeDocsLink) {
          result.repository.docs = fillTemplate(templates.docstemplate);
        }
        if (storeIssuesLink) {
          result.repository.issues = fillTemplate(templates.bugstemplate);
        }
        return result;
      }
    }
  }
  return undefined;
};
