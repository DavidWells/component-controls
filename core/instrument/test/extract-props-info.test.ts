import * as path from 'path';
import { propsInfo } from '../src/misc/propsInfo';

describe('extract-props-info', () => {
  const loadTestFiles = (componentName: string, fileName: string) => {
    it(componentName, async () => {
      expect(
        await propsInfo(
          [
            {
              name: '@component-controls/react-docgen-info',
              test: /\.(js|jsx)$/,
            },
            {
              name: '@component-controls/react-docgen-typescript-info',
              test: /\.(ts|tsx)$/,
            },
          ],
          fileName,
          componentName,
        ),
      ).toMatchSnapshot();
    });
  };
  loadTestFiles(
    'PropsTable',
    path.resolve(__dirname, '../../../ui/blocks/src/PropsTable/PropsTable.tsx'),
  );

  loadTestFiles(
    'Button',
    path.resolve(
      __dirname,
      '../../../node_modules/@theme-ui/components/src/Button.js',
    ),
  );
});
