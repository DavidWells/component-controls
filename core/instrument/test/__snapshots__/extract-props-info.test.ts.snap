// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`extract-props-info Button 1`] = `
Object {
  "description": "",
  "displayName": "Button",
  "methods": Array [],
}
`;

exports[`extract-props-info PropsTable 1`] = `
Object {
  "description": "",
  "displayName": "PropsTable",
  "methods": Array [],
  "props": Object {
    "actions": Object {
      "description": "additional actions provided to the component",
      "parentName": "BlockContainerProps",
      "type": Object {
        "name": "array",
        "raw": "ActionItem[]",
        "value": Array [
          Object {
            "name": "ActionItem",
          },
        ],
      },
    },
    "of": Object {
      "description": "component",
      "parentName": "ComponentInputProps",
      "type": Object {
        "name": "any",
      },
    },
    "title": Object {
      "description": "optional section title for the block",
      "parentName": "BlockContainerProps",
      "type": Object {
        "name": "string",
      },
    },
  },
}
`;
